src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,281 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
 x.2.some
 
 /--
 Every element in the degree zero part of `Aₓ` can be written as `a/x^n` for some `a` and `n : ℕ`,
 `degree_zero_part.deg` picks the numerator `a`
 -/
-def degree_zero_part.num {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) : A :=
+def degree_zero_part.num {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) : A :=
 x.2.some_spec.some.1
 
-lemma degree_zero_part.num_mem {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) :
-  degree_zero_part.num f_deg x ∈ 𝒜 (m * degree_zero_part.deg f_deg x) :=
+lemma degree_zero_part.num_mem {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) :
+  degree_zero_part.num x ∈ 𝒜 (m * degree_zero_part.deg x) :=
 x.2.some_spec.some.2
 
-lemma degree_zero_part.eq {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) :
-  x.1 = mk (degree_zero_part.num f_deg x) ⟨f^(degree_zero_part.deg f_deg x), ⟨_, rfl⟩⟩ :=
+lemma degree_zero_part.eq {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) :
+  x.1 = mk (degree_zero_part.num x) ⟨f^(degree_zero_part.deg x), ⟨_, rfl⟩⟩ :=
 x.2.some_spec.some_spec
 
 lemma degree_zero_part.mul_val {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x y : A⁰_ f_deg) :
   (x * y).1 = x.1 * y.1 := rfl
 
 end
 
+section clear_denominator
+
+-- this is a wrapper around `is_localization.exist_integer_multiples_of_finset`, the main purpose
+-- of this lemma is to make the degree of denominator explicit.
+private lemma clear_denominator {f : A} (s : finset (away f)) :
+  ∃ (n : ℕ), ∀ (x : away f), x ∈ s →
+    x * (mk (f^n) 1 : away f) ∈
+    (λ y, (mk y 1 : localization.away f)) '' set.univ :=
+begin
+  rcases is_localization.exist_integer_multiples_of_finset (submonoid.powers f) s with
+    ⟨⟨_, ⟨n, rfl⟩⟩, h⟩,
+  refine ⟨n, λ x hx, _⟩,
+  rcases h x hx with ⟨a, eq1⟩,
+  induction x using localization.induction_on with data,
+  rcases data with ⟨x, y⟩,
+  change mk a 1 = f^n • _ at eq1,
+  unfold has_scalar.smul localization.smul at eq1,
+  rw [localization.lift_on_mk, smul_eq_mul] at eq1,
+  rw [mk_mul, mul_one, mul_comm, ← eq1],
+  refine ⟨a, trivial, rfl⟩,
+end
+
+end clear_denominator
+
+namespace Top_component
+
+/-
+This section is to construct the homeomorphism between `Proj` restricted at basic open set at
+a homogeneous element `x` and `Spec A⁰ₓ` where `A⁰ₓ` is the degree zero part of the localized
+ring `Aₓ`.
+-/
+
+namespace forward
+
+open ideal
+
+-- This section is to construct the forward direction :
+-- So for any `x` in `Proj| (pbo f)`, we need some point in `Spec A⁰_f`, i.e. a prime ideal,
+-- and we need this correspondence to be continuous in their Zariski topology.
+
+variables {𝒜} {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : Proj| (pbo f))
+
+/--For any `x` in `Proj| (pbo f)`, the corresponding ideal in `Spec A⁰_f`. This fact that this ideal
+is prime is proven in `Top_component.forward.to_fun`-/
+def carrier : ideal (A⁰_ f_deg) :=
+ideal.comap (algebra_map (A⁰_ f_deg) (away f))
+  (ideal.span { y | ∃ (g : A), g ∈ x.1.as_homogeneous_ideal.1 ∧ y = (mk g 1 : away f) })
```

Comment:
Is there a particular reason why this is stated in terms of `mk` instead of `algebra_map`?
IMO the latter (being a ring hom) has better properties and lemmas.

For example, the proof of `men_carrier.clear_denominator` could be golfed into the following, and it avoids the usage of the helper lemma above.

```lean


/--For any `x` in `Proj| (pbo f)`, the corresponding ideal in `Spec A⁰_f`. This fact that this ideal
is prime is proven in `Top_component.forward.to_fun`-/
def carrier : ideal (A⁰_ f_deg) :=
ideal.comap (algebra_map (A⁰_ f_deg) (away f))
  (ideal.span $ algebra_map A (away f) '' x.1.as_homogeneous_ideal)

lemma mem_carrier_iff (z : A⁰_ f_deg) :
  z ∈ carrier f_deg x ↔
  z.1 ∈ ideal.span (algebra_map A (away f) '' x.1.as_homogeneous_ideal) :=
iff.rfl

lemma mem_carrier.clear_denominator [decidable_eq (away f)]
  {z : A⁰_ f_deg} (hz : z ∈ carrier f_deg x) :
  ∃ (c : algebra_map A (away f) '' x.1.as_homogeneous_ideal →₀ away f)
    (N : ℕ)
    (acd : Π y ∈ c.support.image c, A),
    f ^ N • z.1 =
    algebra_map A (away f) (∑ i in c.support.attach,
      acd (c i) (finset.mem_image.mpr ⟨i, ⟨i.2, rfl⟩⟩) * classical.some i.1.2) :=
begin
  erw [mem_carrier_iff, ←submodule_span_eq, finsupp.span_eq_range_total, set.mem_range] at hz,
  rcases hz with ⟨c, eq1⟩,
  rw [finsupp.total_apply, finsupp.sum] at eq1,
  obtain ⟨⟨_, N, rfl⟩, hN⟩ := is_localization.exist_integer_multiples_of_finset (submonoid.powers f)
    (c.support.image c),
  choose acd hacd using hN,
  have prop1 : ∀ i, i ∈ c.support → c i ∈ finset.image c c.support,
  { intros i hi, rw finset.mem_image, refine ⟨_, hi, rfl⟩, },

  refine ⟨c, N, acd, _⟩,
  rw [← eq1, smul_sum, map_sum, ← sum_attach],
  congr' 1,
  ext i,
  rw [_root_.map_mul, hacd, (classical.some_spec i.1.2).2, smul_eq_mul, smul_mul_assoc],
  refl
end
```

Reply:
I used `mk _ _` to calculate with fractions everywhere, especially when latter proving that maps are well defined, backward and forward direction composes to identity etc. Lemmas like `add_mk` `mk_mul` are very easy to use. But If there is any reason to avoid this, I will change it immediately @erdOne.

Reply:
I meant that changing `mk x 1` to `algebra_map R _ x` here would make both the statement and the proof here more concise and easier to read. If this does not cause major issues down the road, I'd argue this is the better way to go.